### ISSUE TITLE
remove automatic type promotion

### ISSUE THREAD

**Issue by lutzhamel** [asteroid-lang/asteroid#241]
There is a fundamental issue in Asteroid where some computations work intuitively and others don't. Consider the following scenarios,
```
let x = 1 + 3.1.
```
This computation is successful by virtue of promoting the integer 1 to the real value 1.0 and then performing a floating point value addition.

However, in the next scenario the computation is not successful,
```
function add_real with (a:%real, b:%real) do
   a+b
end

let x = add_real (1, 3.1).  <<< error
```
Asteroid does not support type promotion during pattern matching, therefore, the above fails since the function expects a real value but is passed an integer in the first component of the pair.

This inconsistent behavior that in some instances type promotion is applied and in others not is confusing to the user.  Therefore I think it is best to remove automatic type promotion all together.  

PS:  I wonder if the Rust folks discovered the same thing and therefore did not implement automatic type promotion in Rust.

**Comment by lutzhamel** [comment]
Should this include `map2boolean`? or should we insist that logical expressions and loop conditions are expressed in true boolean constants?

**Comment by ijchen** [comment]
My $0.02 on `map2boolean` is that logical expressions and loop conditions should be consistent with operators regarding implicit promotion - if `2 + 3.2` doesn't work, neither should `if 3 ...`.

To me, boolean <-> numeric casts seem unlike numeric <-> numeric casts in that they have a fundamentally different meanings - a bool means something is true or false, whereas different numeric types are just different types for expressing the same concept (numbers)

It's sort of nonsensical to have a number in place of a condition, so I've found in my experience that anytime I've done so, it's been a logical error I was glad the compiler or interpreter caught (or frustrated that it didn't). I can't think of a situation where implicit conversion to bools would be helpful - the only case I can think of is possibly treating empty lists as falsy, but even then I prefer the clarity and explicitness of `@length() == 0`

**Comment by lutzhamel** [comment]
@ijchen I agree with you and the current experimental version 2.0.1dev **does not** support pythonic truthiness and therefore supports the explicit programming style along your line of thinking above.

**Comment by lutzhamel** [comment]
resolved with #268
